core/src/main/java/org/apache/calcite/rel/rules/MinusToFilterRule.java

```diff
@@ -0,0 +1,137 @@
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Minus;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.tools.RelBuilder;
+
+import org.immutables.value.Value;
+
+/**
+ * Rule that replaces {@link org.apache.calcite.rel.core.Minus}
+ * operator with {@link org.apache.calcite.rel.core.Filter}
+ * when both inputs are from the same source with only filter conditions differing.
+ *
+ * <p>Example transformation:
+ * <blockquote><pre>
+ * SELECT mgr, comm FROM emp WHERE mgr = 12
+ * EXCEPT
+ * SELECT mgr, comm FROM emp WHERE comm = 5
+ *
+ * to
+ *
+ * SELECT DISTINCT mgr, comm FROM emp
+ * WHERE mgr = 12 AND NOT(comm = 5)
+ * </pre></blockquote>
+ */
+@Value.Enclosing
+public class MinusToFilterRule
+    extends RelRule<MinusToFilterRule.Config>
+    implements TransformationRule {
+
+  /** Creates an MinusToFilterRule. */
+  protected MinusToFilterRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Minus minus = call.rel(0);
+    if (minus.all || minus.getInputs().size() != 2) {
+      return;
+    }
+
+    final RelNode leftInput = minus.getInput(0);
+    final RelNode rightInput = minus.getInput(1);
+
+    final RelBuilder builder = call.builder();
+
+    final RelNode leftBase = getBaseSource(leftInput);
+    final RelNode rightBase = getBaseSource(rightInput);
+
+    RexNode rightCond = extractFilterCond(rightInput);
+    if (!RexUtil.isDeterministic(rightCond)) {
+      return;
+    }
+
+    // Verify they come from the same source
+    if (!leftBase.equals(rightBase)) {
+      return;
+    }
+
+    final RexNode finalCond =
+        builder.and(extractFilterCond(leftInput), builder.not(rightCond));
+
+    builder.push(leftBase)
+        .filter(finalCond)
+        .distinct();
+
+    call.transformTo(builder.build());
+  }
+
+  /**
+   * Returns the base source of the given node,
+   * skipping first filter.
+   */
+  private static RelNode getBaseSource(RelNode node) {
+    RelNode stripped = node.stripped();
+    if (stripped instanceof Filter) {
+      return ((Filter) stripped).getInput();
+    } else {
+      return node;
+    }
+  }
+
+  private static RexNode extractFilterCond(RelNode node) {
+    RelNode stripped = node.stripped();
+    if (stripped instanceof Filter) {
+      return ((Filter) stripped).getCondition();
+    }
+    return node.getCluster().getRexBuilder().makeLiteral(true);
+  }
+
+  /** Rule configuration. */
+  @Value.Immutable
+  public interface Config extends RelRule.Config {
+    Config DEFAULT = ImmutableMinusToFilterRule.Config.of()
+        .withOperandFor(Minus.class);
+
+    @Override default MinusToFilterRule toRule() {
+      return new MinusToFilterRule(this);
+    }
+
+    /** Limit how much complexity can increase during merging.
```

Comment:
I am not sure this is necessary here, because the result is always linear in the size of the input.

Reply:
Done!